hwmv2: convert SiFive and RISC-V QEMU boards

fkokosinski opened the pull request:
Targets in question:
* `hifive1`
* `hifive1_revb`
* `hifive_unleashed`
* `hifive_unmatched`
* `qemu_riscv32_xip`
* `sparkfun_red_v_things_plus`

Converted to:
```
west build -p -b hifive1 samples/hello_world -t run
west build -p -b hifive1@B samples/hello_world -t run
west build -p -b hifive_unleashed samples/hello_world -t run
west build -p -b hifive_unmatched samples/hello_world -t run
west build -p -b sparkfun_red_v_things_plus samples/hello_world
west build -p -b qemu_riscv32/fe310/xip  samples/hello_world -t run
```

Edit:
~~I see that two targets also use SoCs from the FE300 SoC series: `qemu_riscv32_xip` and `sparkfun_red_v_things_plus`. I'll start porting them right away.~~ Done.

fkokosinski committed:
soc/sifive/sifive_freedom: add SiFive Freedom E310 SoC

This commit adds support for the SiFive Freedom E310 SoC for the Zephyr
Hardware Model v2.

Signed-off-by: Filip Kokosinski <fkokosinski@antmicro.com>

fkokosinski committed:
boards: convert SiFive HiFive1 to Zephyr HWMv2

This commit converts the SiFive HiFive1 board (`hifive1` target) to the
Zephyr Hardware Model v2.

Signed-off-by: Filip Kokosinski <fkokosinski@antmicro.com>

fkokosinski committed:
soc/sifive/sifive_freedom: add SiFive Freedom U540 SoC

This commit adds support for the SiFive Freedom U540 SoC for the Zephyr
Hardware Model v2.

Signed-off-by: Filip Kokosinski <fkokosinski@antmicro.com>

fkokosinski committed:
boards: convert SiFive HiFive Unleashed to Zephyr HWMv2

This commit converts the SiFive HiFive Unleashed board (`hifive_unleashed`
target) to the Zephyr Hardware Model v2.

Signed-off-by: Filip Kokosinski <fkokosinski@antmicro.com>

fkokosinski committed:
soc/sifive/sifive_freedom: add SiFive Freedom FU740 SoC

This commit adds support for the SiFive Freedom U740 SoC for the Zephyr
Hardware Model v2.

Signed-off-by: Filip Kokosinski <fkokosinski@antmicro.com>

fkokosinski committed:
boards: convert SiFive HiFive Unmatched to Zephyr HWMv2

This commit converts the SiFive HiFive Unmatched board (`hifive_unmatched`
target) to the Zephyr Hardware Model v2.

Signed-off-by: Filip Kokosinski <fkokosinski@antmicro.com>

fkokosinski committed:
boards: convert SparkFun RED-V Things Plus to Zephyr HWMv2

This commit converts the SparkFun RED-V Things Plus board
(`sparkfun_red_v_things_plus` target) to the Zephyr Hardware Model v2.

Signed-off-by: Filip Kokosinski <fkokosinski@antmicro.com>

nordicjm commented:
Ha, just as I had finished porting hifive1 and revb!

nordicjm commented on soc/sifive/sifive_freedom/CMakeLists.txt:
```diff
@@ -0,0 +1,5 @@
+# Copyright (c) 2024 Nordic Semiconductor
+# SPDX-License-Identifier: Apache-2.0
+
```
v1 files should be removed in the same commit

fkokosinski replied:
Should be removed in the same commit now

nordicjm commented on boards/sifive/hifive1/Kconfig.defconfig:
```diff
@@ -0,0 +1,17 @@
+# SPDX-License-Identifier: Apache-2.0
+
+if BOARD_HIFIVE1
+
```
This is not correct, see https://github.com/nordicjm/zephyr/commits/hwmv2-riscv-1-sifive/ for how I ported it, this configuration in hwmv1 is for the revb board, not the normal board, and the config from the normal board is missing

fkokosinski replied:
Thanks for the link. Should be fixed now

nordicjm commented on boards/sifive/hifive1/Kconfig.hifive1:
```diff
@@ -0,0 +1,5 @@
+# SPDX-License-Identifier: Apache-2.0
+
+config BOARD_HIFIVE1
+	select SOC_SIFIVE_FREEDOM_E310
```
does not select the additional Kconfig for revb

fkokosinski replied:
fixed

nordicjm commented on boards/sifive/hifive1/hifive1.yaml:
```diff
@@ -19,4 +19,3 @@ testing:
     - bluetooth
     - flash
     - newlib
-vendor: sifive
```
??

fkokosinski replied:
Oops... Fixed

nordicjm commented on boards/sifive/hifive1/board.cmake:
```diff
@@ -12,11 +12,18 @@ set(QEMU_FLAGS_${ARCH}
-board_runner_args(openocd --cmd-post-verify "hifive1-post-verify")
-
-include(${ZEPHYR_BASE}/boards/common/openocd.board.cmake)
+if(${BOARD_REVISION} STREQUAL "A")
```
Do not use this, use the Kconfig for the board

nordicjm commented on soc/sifive/sifive_freedom/soc.yml:
```diff
@@ -3,4 +3,4 @@ family:
       - name: fe300
         socs:
-          - name: fe310-g000
+          - name: fe310
```
should be fixed in the original commit

fkokosinski replied:
done

nordicjm commented on soc/sifive/sifive_freedom/fu500/Kconfig.defconfig.u540:
```diff
@@ -2,4 +2,4 @@
 
 config SOC
-	default "u540" if SOC_SIFIVE_FREEDOM_U540
+	default "fu540" if SOC_SIFIVE_FREEDOM_U540
```
??
Should be Kconfig.soc file

fkokosinski replied:
Should be fixed for all SoCs

nordicjm commented on soc/sifive/sifive_freedom/fu700/Kconfig.defconfig:
```diff
@@ -4,7 +4,7 @@
 
 config SOC_SERIES
-	default "u700"
+	default "fu700"
```
Kconfig.soc file

nordicjm commented on soc/sifive/sifive_freedom/fu700/Kconfig.defconfig.u740:
```diff
@@ -2,4 +2,4 @@
 
 config SOC
-	default "u740" if SOC_SIFIVE_FREEDOM_U740
+	default "fu740" if SOC_SIFIVE_FREEDOM_U740
```
Kconfig.soc file

nordicjm commented on boards/sifive/hifive_unmatched/Kconfig.hifive_unmatched:
```diff
@@ -3,4 +3,4 @@
 
 config BOARD_HIFIVE_UNMATCHED
 	bool "HiFive Unmatched target"
-	depends on SOC_SIFIVE_FREEDOM_U740
```
No choice selection for socs or boards

fkokosinski replied:
done

nordicjm commented on boards/sparkfun/sparkfun_red_v_things_plus/board.cmake:
```diff
@@ -1,8 +1,7 @@
 board_runner_args(jlink "--iface=JTAG")
 board_runner_args(jlink "--speed=4000")
 board_runner_args(jlink "--tool-opt=-jtagconf -1,-1")
 board_runner_args(jlink "--tool-opt=-autoconnect 1")
```
Why would the runner file be removed?

fkokosinski replied:
fixed

nordicjm commented on boards/sparkfun/sparkfun_red_v_things_plus/board.cmake:
```diff
@@ -1,8 +1,7 @@
-# Copyright (c) 2019 SiFive Inc.
 # SPDX-License-Identifier: Apache-2.0
```
and the copyright?

fkokosinski replied:
My bad - fixed

nordicjm commented on boards/qemu/qemu_riscv32/qemu_riscv32_fe310_xip.yaml:
```diff
@@ -1,4 +1,4 @@
-identifier: qemu_riscv32_xip
+identifier: qemu_riscv32_fe310_xip
```
```suggestion
identifier: qemu_riscv32/fe310/xip
```

fkokosinski commented on arch/riscv/CMakeLists.txt:
```diff
@@ -9,3 +9,5 @@ set_property(GLOBAL PROPERTY PROPERTY_OUTPUT_FORMAT elf64-littleriscv)
 set_property(GLOBAL PROPERTY PROPERTY_OUTPUT_FORMAT elf32-littleriscv)
 endif()
+
+add_subdirectory_ifdef(CONFIG_RISCV_PRIVILEGED riscv-privileged)
```
Thanks for catching that. Should be fixed now

fkokosinski commented on arch/riscv/Kconfig:
```diff
@@ -366,5 +366,6 @@ config ARCH_HAS_SINGLE_THREAD_SUPPORT
 
 rsource "Kconfig.isa"
 rsource "Kconfig.core"
```
Sure :+1:

nordicjm commented on soc/sifive/sifive_freedom/Kconfig:
```diff
@@ -0,0 +1,7 @@
+# Copyright (c) 2024 Antmicro <www.antmicro.com>
+# SPDX-License-Identifier: Apache-2.0
+
+config SOC_FAMILY_SIFIVE_FREEDOM
```
```suggestion
```

fkokosinski replied:
Done

nordicjm commented on soc/sifive/sifive_freedom/fu500/Kconfig.soc:
```diff
@@ -0,0 +1,16 @@
+	select SOC_FAMILY_SIFIVE_FREEDOM
+
+config SOC_SERIES
+	default "fu500" if SOC_SERIES_SIFIVE_FREEDOM_U500
```
Shouldn't this be U500? Or shouldn't the Kconfig be `SOC_SERIES_SIFIVE_FREEDOM_FU500`?

nordicjm commented on soc/sifive/sifive_freedom/fu700/Kconfig.soc:
```diff
@@ -0,0 +1,16 @@
+	select SOC_FAMILY_SIFIVE_FREEDOM
+
+config SOC_SERIES
+	default "fu700" if SOC_SERIES_SIFIVE_FREEDOM_U700
```
As above, name should match Kconfig

fkokosinski replied:
Dropped the `f` from the start of the string

nordicjm commented on soc/sifive/sifive_freedom/fu700/Kconfig.soc:
```diff
@@ -0,0 +1,16 @@
+	select SOC_SERIES_SIFIVE_FREEDOM_U700
+
+config SOC
+	default "fu740" if SOC_SIFIVE_FREEDOM_U740
```
And here

nordicjm reviewed (commented):
Minor nit.

fkokosinski commented:
Hi @nordicjm, I've addressed your 2nd round of comments. PTAL

nordicjm commented:
Looks good, thanks, should hear back about the privileged riscv part tomorrow

gmarull reviewed (dismissed):
update docs/maintainers as needed

nordicjm commented:
Rebase needed

fkokosinski committed:
MAINTAINERS: update RISC-V arch area paths

This commit updates the paths assigned to the RISC-V area of maintenance to
include targets based on the SiFive Freedom SoC family.

Signed-off-by: Filip Kokosinski <fkokosinski@antmicro.com>

fkokosinski commented:
@gmarull @nordicjm I've updated the docs, rebased and dropped the commit that was moving the `riscv-privileged` dir. PTAL

gmarull reviewed (commented):
IIRC we agreed on not to touch images and merge ignoring the failure to not increase the repo size

fkokosinski commented:
> IIRC we agreed on not to touch images and merge ignoring the failure to not increase the repo size

I've dropped the commit that reduced the image size.

nordicjm commented:
Compliance issues need fixing
Board names in sample/test yaml files need updating

nordicjm commented on boards/sifive/hifive1/Kconfig.hifive1:
```diff
@@ -0,0 +1,6 @@
+
+config BOARD_HIFIVE1
+	select SOC_SIFIVE_FREEDOM_E310
+	select RISCV_CORE_E31 if "$(BOARD_REVISION)" = "B"
```
this line in `Kconfig`

nordicjm replied:
`Kconfig.soc`:
```
config BOARD_HIFIVE1
	select SOC_SIFIVE_FREEDOM_E310
```
`Kconfig`:
```
config BOARD_HIFIVE1
	select RISCV_CORE_E31 if "$(BOARD_REVISION)" = "B"
```

nordicjm commented on soc/sifive/sifive_freedom/Kconfig:
```diff
@@ -0,0 +1,5 @@
+# Copyright (c) 2024 Antmicro <www.antmicro.com>
+# SPDX-License-Identifier: Apache-2.0
+
+rsource "*/Kconfig"
```
```suggestion
if SOC_FAMILY_SIFIVE_FREEDOM

rsource "*/Kconfig"

endif # SOC_FAMILY_SIFIVE_FREEDOM
```

fkokosinski replied:
done

nordicjm commented on boards/boards_legacy/riscv/hifive1_revb/doc/img/hifive1_revb.jpg:
sure you want to delete image? Can keep it if wanted

fkokosinski replied:
I've decided to keep the image for revb

nordicjm reviewed (commented):
Minor nit. left, image comment is optional

fkokosinski committed:
boards: convert SiFive HiFive1 Rev. B to Zephyr HWMv2

This commit converts the SiFive HiFive1 Rev. B board (`hifive1_revb`
target) to the Zephyr Hardware Model v2.

Signed-off-by: Filip Kokosinski <fkokosinski@antmicro.com>

nordicjm commented on soc/sifive/sifive_freedom/CMakeLists.txt:
```diff
@@ -2,4 +2,4 @@
 
 add_subdirectory(common)
-add_subdirectory(${SOC_SERIES})
+add_subdirectory(${CONFIG_SOC_SERIES})
```
this should be `add_subdirectory(${SOC_SERIES})`, if SOC_SERIES is not equal to CONFIG_SOC_SERIES then that's a problem

tejlmand replied:
Hopefully they are always in sync:
https://github.com/zephyrproject-rtos/zephyr/blob/6e0945a82ab0a66728b442e37f7fc9e990e1d00f/cmake/modules/soc_v2.cmake#L26

iirc the `SOC_SERIES` version originates from a time when arch and soc had to be passed to the build, and thus not came through Kconfig.

tejlmand commented on boards/qemu/qemu_riscv32/Kconfig.qemu_riscv32:
```diff
@@ -0,0 +1,7 @@
+
+config BOARD_QEMU_RISCV32_FE310_XIP
+	bool
+	default y
```
lease, no bool, as the `BOARD_<name>` and `BOARD_<full_identifier>` are constructed here:
https://github.com/zephyrproject-rtos/zephyr/blob/db742d03aa8fa469a96435fe85b292e62d061d38/boards/Kconfig.v2#L8
https://github.com/zephyrproject-rtos/zephyr/blob/db742d03aa8fa469a96435fe85b292e62d061d38/boards/Kconfig.v2#L13

Some details can be found here: https://docs.zephyrproject.org/latest/build/kconfig/tips.html#common-kconfig-shorthands

Please do:
```suggestion
config BOARD_QEMU_RISCV32
```
as that will ensure any typos or similar wrong selection is caught by CI.

tejlmand reviewed (commented):
Some minor nits observed.

fkokosinski commented:
Hi @nordicjm, @tejlmand. Thanks for your comments. 

It looks like it will be problematic to just port the `qemu_riscv32/fe310/xip` target right now without porting the rest of RISC-V 32 QEMU targets as well in a single PR without causing any CI fails. For context, only `qemu_riscv32_xip` was using the FE310 SoC; the rest of the `qemu_riscv32` boards were using the generic virt machine. This means I'll have to port the virt `qemu_riscv32`.

I'll start working on it ASAP.

fkokosinski committed:
soc/riscv: convert the QEMU virt RISCV-32 SoC to HWMv2

This commit converts the virt SoC (generic virt machine) to the Zephyr
HWMv2.

Signed-off-by: Filip Kokosinski <fkokosinski@antmicro.com>

fkokosinski committed:
boards: convert QEMU RISC-V 32 bit board to Zephyr HWMv2

This commit converts the QEMU RISCV-V 32 bit board to Zephyr HWMvW. This
includes the following former targets:
* qemu_riscv32
* qemu_riscv32_smp
* qemu_riscv32_xip

Signed-off-by: Filip Kokosinski <fkokosinski@antmicro.com>

fkokosinski committed:
boards: convert QEMU RISC-V 64 bit board to Zephyr HWMv2

This commit converts the QEMU RISCV-V 64 bit board to Zephyr HWMvW. This
includes the following former targets:
* qemu_riscv64
* qemu_riscv64_smp

Signed-off-by: Filip Kokosinski <fkokosinski@antmicro.com>

fkokosinski committed:
boards: convert QEMU RV32E board to Zephyr HWMv2

This commit converts the QEMU RV32E board to Zephyr HWMvW. This includes
the following former target: qemu_riscv32e.

Signed-off-by: Filip Kokosinski <fkokosinski@antmicro.com>

nordicjm commented on soc/sifive/sifive_freedom/e300/Kconfig.soc:
```diff
@@ -0,0 +1,16 @@
+	select SOC_FAMILY_SIFIVE_FREEDOM
+
+config SOC_SERIES
+	default "e300" if SOC_SERIES_SIFIVE_FREEDOM_E300
```
We are planning to have the value and name of this linked in future: https://github.com/zephyrproject-rtos/zephyr/issues/69317
But this raises an interesting issue @tejlmand because if just "e300" is used, that's so generic it's bound to have a name clash in future

nordicjm replied:
It's a future thing, e300 is fine for now but yes in future might have to expand it

tejlmand replied:
> It's a future thing, e300 is fine for now but yes in future might have to expand it

I think current proposal is fine.
If we're later going to use those values internally, like we do the the `CONFIG_SOC`, then we should be generating those.
Just like we do for boards.

But there are a couple of reasons this is not done in first iteration of hwmv2, so let's not worry to much now.
But thanks for raising awareness on this @nordicjm 👍

nordicjm commented on soc/riscv/qemu_virt_riscv/Kconfig:
```diff
@@ -11,3 +10,5 @@ config SOC_RISCV_VIRT
 	select RISCV_PRIVILEGED
 	select RISCV_HAS_PLIC
+
+rsource "*/Kconfig"
```
Needs Kconfig guard

fkokosinski replied:
Thanks, fixed

nordicjm commented on soc/riscv/qemu_virt_riscv/qemu_virt_riscv32/Kconfig:
```diff
@@ -0,0 +1,8 @@
+	select CPU_HAS_FPU
+	select RISCV_ISA_RV32I
+	select RISCV_ISA_EXT_ZICSR
+	select RISCV_ISA_EXT_ZIFENCEI
```
qemu_virt_riscv32 and qemu_virt_riscv64 should probably become one board and use variant to specify which one is being built where no variant = 32-bit and variant "64" = 64-bit, thoughts @gmarull @aescolar ?

fkokosinski replied:
I don't have a strong preference on this. But if we go ahead and convert them to a single board with variants, I suppose that `qemu_virt_riscv32e` should also be included in the process?

tejlmand replied:
initial feeling is that it's the SoC which is either 32 bit or 64bit, and thus you select the SoC of the arch you want.
Thus the board indirectly select 32 / 64 bit based on the SoC it selects.

Not the board which should directly select 32 or 64 bit.

fkokosinski commented:
Btw. the runners seem to be experiencing some problems:
```The self-hosted runner: zephyr-runner-linux-x64-4xlarge-deployment-g2tcl-lwwnp lost communication with the server. Verify the machine is running and has a healthy network connection. Anything in your workflow that terminates the runner process, starves it for CPU/Memory, or blocks its network access can cause this error.```

I'll wait a bit before restarting the workflow to not flood the whole thing with more workload while it's not responding.

fkokosinski commented:
OK, the CI now only reports one error that looks unrelated to this PR:
>tests/subsys/mgmt/mcumgr/os_mgmt_info/mgmt.mcumgr.os.info.bt (nrf9160dk/nrf52840:tests/subsys/mgmt/mcumgr/os_mgmt_info/mgmt.mcumgr.os.info.bt) with error

nordicjm commented on soc/sifive/sifive_freedom/u700/Kconfig:
```diff
@@ -0,0 +1,19 @@
+# Copyright (c) 2024 Antmicro <www.antmicro.com>
+# SPDX-License-Identifier: Apache-2.0
+
+config SOC_SERIES_SIFIVE_FREEDOM_U700
```
this should be FU700

nordicjm commented on soc/sifive/sifive_freedom/u700/Kconfig.soc:
```diff
@@ -0,0 +1,16 @@
+	bool
+	select SOC_FAMILY_SIFIVE_FREEDOM
+
+config SOC_SERIES
+	default "u700" if SOC_SERIES_SIFIVE_FREEDOM_U700
+
+config SOC_SIFIVE_FREEDOM_U740
+	bool
+	select SOC_SERIES_SIFIVE_FREEDOM_U700
+
+config SOC
+	default "u740" if SOC_SIFIVE_FREEDOM_U740
```
these should all be fu7**

nordicjm commented on soc/sifive/sifive_freedom/e300/Kconfig.soc:
```diff
@@ -0,0 +1,16 @@
+config SOC_SERIES
+	default "e300" if SOC_SERIES_SIFIVE_FREEDOM_E300
+
+config SOC_SIFIVE_FREEDOM_E310
+	bool
+	select SOC_SERIES_SIFIVE_FREEDOM_E300
+
+config SOC
+	default "e310" if SOC_SIFIVE_FREEDOM_E310
```
fe3**

nordicjm commented on soc/sifive/sifive_freedom/u500/Kconfig:
```diff
@@ -1,22 +1,20 @@
-choice
-	prompt "SiFive Freedom SOC implementation"
-	depends on SOC_SERIES_SIFIVE_FREEDOM_U500
+config SOC_SERIES_SIFIVE_FREEDOM_U500
```
FU500

nordicjm commented on soc/sifive/sifive_freedom/u500/Kconfig.soc:
```diff
@@ -0,0 +1,16 @@
+	bool
+	select SOC_FAMILY_SIFIVE_FREEDOM
+
+config SOC_SERIES
+	default "u500" if SOC_SERIES_SIFIVE_FREEDOM_U500
+
+config SOC_SIFIVE_FREEDOM_U540
+	bool
+	select SOC_SERIES_SIFIVE_FREEDOM_U500
+
+config SOC
+	default "u540" if SOC_SIFIVE_FREEDOM_U540
```
fu5** and FU5**

nordicjm reviewed (changes requested):
Also rebase will fix the test failure

nordicjm reviewed (approved):
Thanks for port!
